src/samples/audio/smart_amp_test_ipc4.c

```diff
@@ -0,0 +1,438 @@
+			     struct output_stream_buffer *output_buffers, int num_output_buffers)
+{
+	struct smart_amp_data *sad = module_get_private_data(mod);
+	struct comp_dev *dev = mod->dev;
+	struct comp_buffer __sparse_cache *fb_buf_c;
+	struct module_source_info __sparse_cache *mod_source_info;
+	uint32_t avail_passthrough_frames;
+	uint32_t avail_frames = 0;
+	uint32_t sink_bytes;
+
+	mod_source_info = module_source_info_acquire(mod->source_info);
+
+	smart_amp_input_setup(mod, input_buffers, num_input_buffers);
+	avail_passthrough_frames = sad->src_input->size;
+	if (sad->fb_input) {
+		fb_buf_c = attr_container_of(sad->fb_input->data,
+					     struct comp_buffer __sparse_cache,
+					     stream, __sparse_cache);
+		if (fb_buf_c->source && comp_get_state(dev, fb_buf_c->source) == dev->state) {
+			/* feedback */
+			avail_frames = MIN(avail_passthrough_frames,
+					   sad->fb_input->size);
+
+			sad->process(mod, sad->fb_input, &output_buffers[0],
+				     avail_frames, sad->config.feedback_ch_map);
+		}
+	}
+
+	if (!avail_frames)
+		avail_frames = avail_passthrough_frames;
+
+	/* bytes calculation */
+	sink_bytes = avail_frames *
+		audio_stream_frame_bytes(output_buffers[0].data);
+
+	/* process data */
+	sad->process(mod, sad->src_input, &output_buffers[0],
+		     avail_frames, sad->config.source_ch_map);
+
+	output_buffers[0].size = sink_bytes;
+
+	module_source_info_release(mod_source_info);
+	return 0;
+}
+
+static int smart_amp_reset(struct processing_module *mod)
+{
+	struct comp_dev *dev = mod->dev;
+
+	comp_dbg(dev, "smart_amp_reset()");
+
+	return 0;
+}
+
+static int smart_amp_prepare(struct processing_module *mod)
+{
+	struct smart_amp_data *sad = module_get_private_data(mod);
+	struct comp_dev *dev = mod->dev;
+	struct comp_buffer *source_buffer;
+	struct comp_buffer *sink_buffer;
+	struct comp_buffer __sparse_cache *buffer_c;
+	struct list_item *blist;
+	int ret;
+
+	ret = smart_amp_params(mod);
+	if (ret < 0)
+		return ret;
+
+	comp_dbg(dev, "smart_amp_prepare()");
+	/* searching for stream and feedback source buffers */
+	list_for_item(blist, &dev->bsource_list) {
+		source_buffer = container_of(blist, struct comp_buffer,
+					     sink_list);
+		buffer_c = buffer_acquire(source_buffer);
+		audio_stream_init_alignment_constants(1, 1, &buffer_c->stream);
+		if (IPC4_SINK_QUEUE_ID(buffer_c->id) == SOF_SMART_AMP_FEEDBACK_QUEUE_ID) {
+			buffer_c->stream.channels = sad->config.feedback_channels;
+			buffer_c->stream.rate = mod->priv.cfg.base_cfg.audio_fmt.sampling_frequency;
+		}
+		buffer_release(buffer_c);
+	}
+
+	sink_buffer = list_first_item(&dev->bsink_list, struct comp_buffer, source_list);
+	buffer_c = buffer_acquire(sink_buffer);
+	sad->out_channels = audio_stream_get_channels(&buffer_c->stream);
+	audio_stream_init_alignment_constants(1, 1, &buffer_c->stream);
+	sad->process = get_smart_amp_process(dev, buffer_c);
+	buffer_release(buffer_c);
+
+	sad->input_stream_num = 0;
+	sad->fb_input = NULL;
```

Comment:
in prepare function input_stream_num & fb_input can be setup (1) if feedback stream is started before smart amp, input_stream_num = 2 , fb_input = feedback  or (2) input_stream_num = 1, fb_input = NULL. Please check IPC3 path, they are set up here.

Reply:
from the original code and my test, seems while prepare we always only get the source stream. I 'd like to keep my code now to avoid the accident cases if we keep playback path and start-> stop -> start capture path.

Reply:
now we should look up every process since the input buffer will not be persistent later soon.